authorization_code/app.js

```diff
@@ -0,0 +1,147 @@
+/**
+ * This is an example of a basic node.js script that performs
+ * the Authorization Code oAuth2 flow to authenticate against
+ * the Spotify Accounts.
+ *
+ * For more information, read
+ * https://developer.spotify.com/web-api/authorization-guide/#authorization_code_flow
+ */
+
+var express = require('express'); // Express web server framework
+var request = require('request'); // "Request" library
+var cors = require('cors');
+var querystring = require('querystring');
+var cookieParser = require('cookie-parser');
+
+var client_id = 'bba78fd3d447421381da369e658e3f92'; // Your client id
+var client_secret = '0d8c939ee8f34320af774c02f42d96b5'; // Your secret
+var redirect_uri = 'http://localhost:8888/callback'; // Your redirect uri
+
+/**
+ * Generates a random string containing numbers and letters
+ * @param  {number} length The length of the string
+ * @return {string} The generated string
+ */
+var generateRandomString = function(length) {
+  var text = '';
+  var possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+
+  for (var i = 0; i < length; i++) {
+    text += possible.charAt(Math.floor(Math.random() * possible.length));
+  }
+  return text;
+};
+
+var stateKey = 'spotify_auth_state';
+
+var app = express();
+
+app.use(express.static(__dirname + '/public'))
+   .use(cors())
+   .use(cookieParser());
+
+app.get('/login', function(req, res) {
+
+  var state = generateRandomString(16);
+  res.cookie(stateKey, state);
+
+  // your application requests authorization
+  var scope = 'user-read-private user-read-email';
+  res.redirect('https://accounts.spotify.com/authorize?' +
```

Comment:
This is an express command. When the browser receives the http response, it will redirect to `https://accounts.spotify.com/authorize?` but it will also have information encoded at the end of the url.

This will include 
```
      response_type: 'code', // how does the server want to get a response
      client_id: client_id, // unique id awesomeSite.com registers with spotify so spotify knows whos asking.
      scope: scope, // The permissions requested
      redirect_uri: redirect_uri, // this is the url to send a response back to. This will be awesomeSite.com/callback below
      state: state // the unique string we mentioned earlier.
```

Note: `state` and the cookie `spotify_auth_state` are different things, set to the same value. This is a security feature. We don't want someone maliciously pretending to be a visitor of awesomeSite.com and sending a request from a different computer. Later down, we compare `state` send via spotify to `spotify_auth_state` in the browser to make sure its the same computer the whole time.

Reply:
The whole redirect url might look like `https://accounts.spotify.com/authorize?response_type=code&client_id=awesomeSite&scope=user-read-private_user-read-email&redirect_uri=https://awesomeSite.com/callback&state=randomString123`